sympy/core/operations.py

```diff
@@ -504,3 +524,185 @@ def args(self):
+            already sympified, for faster processing.
+        """
+        if sympify:
+            args = tuple(map(_sympify, args))
+        types = frozenset(map(type, args))
+
+        # If sympify=True was passed, no need to sympify again so pass sympify=False.
+        # If sympify=False was passed, all args are already sympified so pass sympify=False.
+        return self.dispatch(types)(*args, sympify=False, **kwargs)
+
+    @cacheit
+    def dispatch(self, types):
+        """
+        Select the type with highest priority, and return its handler.
+        """
+
+        # Quick exit for the case where all arguments are of same type
+        # or all handlermethods are same
+        handlers = set(map(self._handlergetter, types))
+        if len(handlers) == 1:
+            h, = handlers
+            return h
+
+        # Sieve the types with low op_priority or no handler
+        # This makes it faster than using dispatcher on every types.
+        sieved_data = tied_max(
+            types, key=lambda typ: getattr(typ, '_op_priority', 0)
+        )
+
+        # Only one type remains
+        if len(sieved_data) == 1:
+            typ, = sieved_data
+            return self._handlergetter(typ)
+
+        # Recursively select with registered binary priority
+        for i, typ in enumerate(sieved_data):
+            if i == 0:
+                result_type = typ
+            else:
+                dispatched_selector = self._dispatcher.dispatch(result_type, typ)
+                selected_type = dispatched_selector(result_type, typ)
+
+                if not isinstance(selected_type, type):
+                    raise RuntimeError(
+                        "Dispatcher for {!r} and {!r} must return a type, but got {!r}".format(
+                        result_type, typ, selected_type
+                    ))
+
+                handler = self._handlergetter(selected_type)
+                if handler is None:
+                    raise RuntimeError(
+                        "Dispatcher for {!r} and {!r} returned {!r} which does not have {!r} method.".format(
+                        result_type, typ, selected_type, self.handlermethod
+                    ))
+
+                result_type = selected_type
+
+        # return handler method of selected argument
+        return handler
+
+    @property
+    def __doc__(self):
+        docs = ["Multiply dispatched method: %s" % self.name]
+
+        if self.doc:
+            docs.append(self.doc)
+
+        docs.append("Registered priority relations are as follows:")
+
+        other = []
+        for sig in self._dispatcher.ordering[::-1]:
+            func = self._dispatcher.funcs[sig]
+            if func.__doc__:
+                s = 'Inputs: <%s>\n' % str_signature(sig)
+                s += '-' * len(s) + '\n'
+                s += func.__doc__.strip()
+                docs.append(s)
+            else:
+                other.append(str_signature(sig))
```

Comment:
A possible variation would be
```suggestion
        other = []
        for func, sigs in itertools.groupby(self._dispatcher.ordering[::-1], lambda sig: self._dispatcher.funcs[sig]):
            if func.__doc__:
                s = 'Inputs: %s\n' % ' or '.join('<%s>' % str_signature(sig) for sig in sigs))
                s += '-' * len(s) + '\n'
                s += func.__doc__.strip()
                docs.append(s)
            else:
                other += list(sigs)
```
which would group together functions that are the same